### ISSUE TITLE
Cannot use local DNS servers together with exit node

### ISSUE THREAD

**Issue by Zocker1999NET** [tailscale/tailscale#8350]
### What is the issue?

I'm using an always-on exit node on my PCs. With this configuration, it no longer works to use  DNS servers in local ranges (e.g. `10.10.10.10`), even when these are reachable for other tools while Tailscale is running. Instead, they are required to be reachable by the exit node.

I can reproduce this issue in two different setups. Both setups have an external server as exit node in common, which cannot reach the local DNS server due to security. That server does not need to reach this server, as this server does not use Tailscale for DNS. The setups differ in these configurations:

1. My PC is stationary and is in the same local network as the local DNS. It has `--accept-routes=false` for less delay (see #1227). So it may reach the local DNS directly.
2. My laptop is mobile and so has `--accept-routes=yes` to ensure it can always reach my local setups, including the DNS. With the route, it can reach the local DNS as well.

Disabling the exit node immediately restores the functionality in both setups.

### Steps to reproduce

For simplification, the steps assume a mobile setup.

1. Setup a local DNS server with custom entries.
2. Setup an exit node, which may not reach the local DNS (different network, no Tailscale routes or deny via ACL).
3. Setup a router in the same network as the local DNS server, so it may reach it directly. Fully configure the local subnet to be reachable by Tailscale routes.
4. Configure in Tailscale's Web UI a restricted nameserver for the custom entries, using the local DNS's IP.
5. Setup "a simple client" with dns, exit node & routes.
6. Verify the client may reach the DNS directly (ping, dig, …)
7. Check that the Tailscale's DNS server (`100.100.100.100`) times out when requesting a custom entry.

### Are there any recent changes that introduced the issue?

This definitely worked in an older version, however, I cannot determine which version introduced this issue, as I haven't relied on this feature on a daily basis.

### OS

Linux

### OS version

Debian testing/unstable

### Tailscale version

1.42.0

### Other software

systemd-nameserver is configured in front of Tailscale DNS for mDNS and other stuff, but I've verified that the Tailscale DNS times out by directly asking it by using `dig custom.mydomain @100.100.100.100`.

### Bug report

BUG-61ed798e7e05dd0ab5418968cc9693edc8b54953071b2f018aac499dabadd086-20230614210315Z-36c3d1a05929f1aa

**Comment by DentonGentry** [comment]
The intent with an exit node is that the device, as fully as possible, teleport to the location of the exit node. DNS queries are forwarded to the exit node in order to get the geoIP response of being at the location of the exit node.

May I ask the intent of using the exit node?

**Comment by Zocker1999NET** [comment]
Yes, the intent is to route all of my traffic through one well-working & trusted ISP where this exit node lives.
At home, I primarily use this setup to increase stability (TL;DR: My home ISP has a good peering to the other ISP, but not really to the services I use. Routing everything through my server increases latency by a bit, but makes the connection overall faster & more stable).
On the go, it allows me to use public connections without worrying about my privacy.
Additionally, as all services see only one IP for all my devices, I can easily interpret notifications like "someone logged in as IP X".

In the case of the DNS being a public one, I understand why it makes sense to route all DNS requests via the exit node. In my case, where those are private IPs, which are configured to use another route, it feels counter-intuitive to me. Could there maybe be an exemption for private IPs and subnets served by subnet routes?

**Comment by abuchanan920** [comment]
Here is a use case to consider (that I just ran into)...

I have an beefy Ubuntu box that is my exit node/subnet router for my home network AND my internal DNS server AND serves various other web services.

If I configure a mobile client to use the exit node, I get the DNS response that the exit node would get for its OWN name (as validated by dig in both instances).

Alas that response seems to come from the exit node resolver and is...127.0.1.1. This is because the exit node resolver checks its /etc/hosts file first, and that entry is pretty much always going to be there. I think it even gets added back by something if you try to remove it.

As you can imagine, that isn't terribly useful for the client because when I try to connect via that IP it doesn't work. It DOES from the exit node directly, so something isn't working exactly the same as if the traffic came from the exit node to itself. I suspect localhost traffic is special cased on the client side, as I don't see anything if I tcpdump on the server and try to access it from the client when it is using that address.

In any case, the simplest solution for me would be if I had some way to opt out of using the exit node resolver and continued to use my client's DNS resolver instead. That is set up to use the same DNS server that the exit node is using anyway, and gets the proper routable  internal IP address of that server rather than the localhost one (as tested with dig when the exit node setting is turned off) and would still get me the benefits you described above.

BTW - Thanks for the great tool!
